redash/tasks/failure_report.py

```diff
@@ -0,0 +1,65 @@
+import datetime
+from collections import Counter
+from flask import render_template
+from redash.tasks.general import send_mail
+from redash.worker import celery
+from redash import redis_connection, settings
+from redash.utils import json_dumps, json_loads, base_url
+
+
+@celery.task(name="redash.tasks.send_aggregated_errors")
+def send_aggregated_errors(email_address):
+    key = 'aggregated_failures:{}'.format(email_address)
+    errors = [json_loads(e) for e in redis_connection.lrange(key, 0, -1)]
+
+    if errors:
+        errors.reverse()
+        occurrences = Counter((e.get('id'), e.get('message')) for e in errors)
+        unique_errors = {(e.get('id'), e.get('message')): e for e in errors}
+
+        context = {
+            'failures': [{
+                    'base_url': v.get('base_url'),
+                    'id': v.get('id'),
+                    'name': v.get('name'),
+                    'failed_at': v.get('failed_at'),
+                    'failure_reason': v.get('message'),
+                    'failure_count': occurrences[k],
+                    'comment': v.get('comment')
+            } for k, v in unique_errors.iteritems()]
+        }
+
+        html = render_template('emails/failures.html', **context)
+        text = render_template('emails/failures.txt', **context)
+        subject = "Redash failed to execute {} of your queries".format(len(unique_errors.keys()))
+        send_mail.delay([email_address], subject, html, text)
+
+    redis_connection.delete(key)
+
+
+def notify_of_failure(message, query):
+    if not settings.organization.SEND_EMAIL_ON_FAILED_SCHEDULED_QUERIES:
+        return
+
+    if query.schedule_failures < settings.MAX_FAILURE_REPORTS_PER_QUERY:
+        key = 'aggregated_failures:{}'.format(query.user.email)
+        reporting_will_soon_stop = query.schedule_failures > settings.MAX_FAILURE_REPORTS_PER_QUERY * 0.75
+        comment = """NOTICE: This query has failed a total of {failure_count} times.
+                     Reporting may stop when the query exceeds {max_failure_reports} overall failures.""".format(
+            failure_count=query.schedule_failures,
+            max_failure_reports=settings.MAX_FAILURE_REPORTS_PER_QUERY
+        ) if reporting_will_soon_stop else ''
+
+        redis_connection.lpush(key, json_dumps({
+            'id': query.id,
+            'name': query.name,
+            'base_url': base_url(query.org),
+            'message': message,
+            'comment': comment,
+            'failed_at': datetime.datetime.utcnow().strftime("%B %d, %Y %I:%M%p UTC")
+        }))
+
+        if not redis_connection.exists('{}:pending'.format(key)):
+            send_aggregated_errors.apply_async(args=(query.user.email,), countdown=settings.SEND_FAILURE_EMAIL_INTERVAL)
```

Comment:
I would avoid using Celery features like `countdown` as it can result in multiple executions of the same task 😞 Iirc, if `SEND_FAILURE_EMAIL_INTERVAL` > Celery's `VISIBILITY_TIMEOUT` (1 hour by default), you will get this task to execute more than once.

We can have a periodic task run every X minutes, that will check if there are any reports to send.

Reply:
This makes things simpler! I guess I needed an outside POV to realize that it really doesn't matter if the report is sent 60 minutes _after_ the failure, since these are scheduled queries...

Anyway, this is changed in 2fb4882 and it allowed me to remove the pending keys 6f0066e.

Reply:
I think the benefit of sending it after 60 minutes is that it's more likely to pick up more failing queries in a single email, no?

Reply:
Yeah, but only for the first report. IMO it simplifies the code and less likely to cause duplication issues so I'm fine with it as it is now.

Reply:
I think that with a 60 seconds window the chances of it actually aggregating anything are slim. Maybe we should increase the default to 5 minutes

Reply:
It's [60 minutes](https://github.com/getredash/redash/pull/3798/files#diff-ad93995bb355251c02f87fc02db69351R44).

Reply:
:man_facepalming: